src/gadgets/k_mix.rs

```diff
@@ -34,23 +34,33 @@ pub fn fill_cs<CS: ConstraintSystem>(
     Ok((mix_in, mix_out))
 }
 
+// Calls `k` mix gadgets, using mix_in and mix_mid as inputs, and mix_mid and mix_out as outputs.
 fn call_mix_gadget<CS: ConstraintSystem>(
     cs: &mut CS,
     mix_in: &Vec<AllocatedValue>,
     mix_mid: &Vec<AllocatedValue>,
     mix_out: &Vec<AllocatedValue>,
 ) -> Result<(), R1CSError> {
+    let k = mix_out.len() - 1;
+    if mix_in.len() != k + 1 || mix_mid.len() != k - 1 {
+        return Err(R1CSError::GadgetError {
+            description: "Lengths of inputs are incorrect for call_mix_gadget in k_mix".to_string(),
+        });
+    }
+
+    // The first value of mix_in, to prepend to mix_mid for creating A inputs.
     let first_in = mix_in[0].clone();
+    // The last value of mix_out, to append to mix_mid for creating D outputs.
     let last_out = mix_out[mix_out.len() - 1].clone();
 
-    // For each 2-mix, constrain A, B, C, D:
+    // For each of the `k` mix gadget calls, constrain A, B, C, D (where i is in 0..k):
```

Comment:
wait, the loop seems to go k-1 times, not k times. So it's `0..(k-1)` in rust's exclusive-end range notation (`x..y` == `[x,y)`). Or am i confused?

Reply:
i'd avoid using the range notation entirely :-)

Reply:
my bad... I was using `k` as the number of mix gadgets, when really `k` should be the input length and we should have `k-1` mix gadgets. fixed now!